### ISSUE TITLE
Getting NaNs when fine-tuning after the first epoch

### ISSUE THREAD

**Issue by ahrm** [cactus-compute/needle#53]
When I try to run
```
needle finetune data.jsonl --epochs 3 --checkpoint needle2.pkl
```

I get something like this:

```
checkpoints/needle2.pkl not found locally, downloading from HF...
training on 200 examples, seq_len 1024
LoRA rank 16 on 10 weight groups (compiling...)
epoch 1/3  step 1/39  loss 1.1031
epoch 1/3  step 2/39  loss nan
```

This happens both on mac and on google colab. Also it does not depend on the learning rate, I have tried it with very low learning rates with the same result. (this is using version 2.0.1)

**Comment by HenryNdubuaku** [comment]
@jakmro could you have a look?

**Comment by vishnoorath** [comment]
I had the same issue : LLm answer
Found the root cause. Two problems:

**Wrong LoRA targets** — LoRA is targeting mtp_block attention projections which are all zeros (range [-0, 0]). The actual model weights with real values live in stack.layers.block.self_attn. LoRA is training dead weights.

**float16 precision** — LoRA matrices get cast to float16, causing gradient overflow (NaN after step 1).

C:\Users\VISHNOORATH\AppData\Local\hermes\hermes-agent\venv\Lib\site-packages\needle\model\finetune.py
@@ -247,8 +247,16 @@
 
```
 def lora_target_paths(params):
     from flax.traverse_util import flatten_dict
-    return [path for path in flatten_dict(params)
-            if path[-1] == "kernel" and any(t in path for t in LORA_TARGETS)]
+    flat = flatten_dict(params)
+    # Target the actual stack layers (not mtp_block which has zeroed weights)
+    paths = [path for path in flat
+             if path[-1] == "kernel"
+             and "stack" in path
+             and "layers" in path
+             and any(t in path for t in LORA_TARGETS]
+    # Filter to only weights with non-zero magnitude (skip dead weights)
+    paths = [p for p in paths if jnp.max(jnp.abs(flat[p])) > 1e-6]
+    return paths
```
 
 ```
 def lora_target_paths(params):
+    import jax.numpy as jnp
     from flax.traverse_util import flatten_dict
     flat = flatten_dict(params)
     # Target the actual stack layers (not mtp_block which has zeroed weights)
         in_dim, out_dim = weight.shape[-2], weight.shape[-1]
         lead = weight.shape[:-2]
         key, sub = jax.random.split(key)
+        # Always use float32 for LoRA matrices to avoid float16 overflow -> NaN
         lora[path] = {
-            "A": (jax.random.normal(sub, lead + (in_dim, rank), jnp.float32) / rank).astype(weight.dtype),
-            "B": jnp.zeros(lead + (rank, out_dim), weight.dtype),
+            "A": jax.random.normal(sub, lead + (in_dim, rank), jnp.float32) / rank,
+            "B": jnp.zeros(lead + (rank, out_dim), jnp.float32),
         }
     return lora
 
@@ -282,7 +284,9 @@
     from flax.traverse_util import flatten_dict, unflatten_dict
     flat = dict(flatten_dict(params))
     for path, adapter in lora.items():
-        flat[path] = flat[path] + scale * jnp.matmul(adapter["A"], adapter["B"]).astype(flat[path].dtype)
+        orig = flat[path]
+        delta = jnp.matmul(adapter["A"], adapter["B"])  # float32
+        flat[path] = orig + (scale * delta).astype(orig.dtype)
     return unflatten_dict(flat)
```

**Comment by tchivs** [comment]
Independent reproduction + verification of the two root causes (adding precision + numeric evidence).

**1. dtype is bf16, not fp16.** The model's default dtype is `jnp.bfloat16` (`needle/model/architecture.py`), so `init_lora` casts LoRA A/B via `astype(weight.dtype)` to **bfloat16**, not float16. bf16 matmul overflows on CPU (loss -> NaN from step 2); verified on Windows x64 CPU with both 2.0.1 and 2.0.2 packages, at lr 1e-5..1e-4.

**2. Dead-weight LoRA targets confirmed numerically.** From `checkpoints/needle2.pkl`:

```
mtp_block/self_attn proj kernels: max_abs = 1.67e-06   (dead, ~0)
stack/layers/.../self_attn proj kernels: max_abs = 22.97 (real weights)
```

So 5 of the 10 LoRA groups train zeroed `mtp_block` weights.

**Fix verified end-to-end** (pristine 2.0.2 finetune.py + the three changes: stack/layers filter with magnitude guard, fp32 LoRA init, fp32 merge matmul):
```
epoch 1/6 loss 3.70 → epoch 6/6 loss 3.73   (no NaN at any step)
```
Converges on 4 training samples (compound zh-CN UI examples) on CPU-only JAX.

One remaining blocker for deployment: `needle build` output tag `0x05E12A82` vs native engine expecting `0x05E12A83` (see #58) — the engine and build tool appear out of sync.

**Comment by HenryNdubuaku** [comment]
This has now been fixed and merged, rolling out with the new release today

**Comment by etchevarneM** [comment]
Hit this independently while evaluating Needle for a tool-routing use case, isolated it down to a 2-example toy dataset (loss finite at step 1, 
NaN from step 2 onward, reproducible regardless of learning rate, sequence length, or dataset content) before finding this was already reported. 
Confirmed the merged fix resolves it cleanly. Thanks for the quick turnaround!
